### DISCUSSION TITLE
JARVIS Technologies & Format

### DISCUSSION THREAD

**Discussion by MRita443** [FEUP-MEIC-DS-2024-25/ai4sd#10] [Ideas]
Hi guys! I want to transpose a feature we developed for the prototype to Jarvis. We developed interaction with the GitHub API to fetch a (for now public) repository's Python files, given its URL. We did this by making our assistant into a GitHub App, and authenticating it in GitHub API as an app installation ([more info on authentication methods here](https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28)). We developed this in Python.

I was thinking of starting out JARVIS with this as a base, making this service available on the Echoes. This would involve making JARVIS a GitHub App, though. I was also thinking of adapting this perhaps to Node.js, because most of the GitHub API Docs are in JS. What do you think?

**Comment by Af0ns0836** [comment]
I think it's a good idea, because our group wanted to do a similar thing. We wanted to fetch the documentation and git commits from a repository, given its URL. I believe we can add this feature to JARVIS. Maybe fetch the entire repository or choose only the things we need?

**Comment by davehubber** [comment]
I think that if the idea is for JARVIS to poll used repositories and keep them updated, then we only need 1 function for fetching, that fetches the entire repository. The repository will then be stored and updated on Nexus, and super-heroes access the repository through Nexus directly, for the contents (python files, logs, commits, images) that they might need.
The first request is much more weighty, but it sort of future proofs subsequent requests.

So that the assistant that makes the request for a repository that is not on Nexus, doesn't wait longer than what it needs, we could try and figure out a way for it to return what the assistant is requiring as soon as possible, and then asynchronously finish extracting the repository, although that seems a bit complex for now, and I'm not even sure if this wait time is going to be much of an issue...

What do you think?

**Reply by davehubber** [reply]
Okay, actually never mind 😄
Considering that we also want to store commits, logs e.t.c, keeping the entire repository seems excessive. I think we could organize a folder per repository in Nexus yes, but then have subfolders for each of the things the assistants request (we know what they are and they won't change). Have a folder for commits, logs, code files (could be further organized by language), images... Jarvis fetches and polls that and only that. Subfolders are created for a repo upon the first fetch request of that category, made by a superhero.
I think this is probably the best approach
